### PR TITLE
add GLES 2 renderer for 2D

### PR DESCRIPTION
**Pull request by karroffel**
This commit adds a new rendering backend, GLES2, and adds a
project setting to enable it.

Currently this backend can only be used on the X11 platform,
but integrating into other platforms is planned.

### CODE REVIEW

**Comment by groud** [comment]
How about Vulkan support ?

Just kidding, this is a great job ! :D

**Comment by groud** [comment]
However, the static checks do not pass :/

**Comment by groud** [comment]
Could you squash the commits together too ? XD

**Comment by karroffel** [comment]
@groud I can, and in fact the initial commit here is actually just 34 commits squashed together, just let me make the CI happy and then I'll do it :D

**Comment by groud** [comment]
Well, it was just to avoid to making Travis run twice. 
But as it works fine those days I guess it's ok. :)

**Comment by karroffel** [comment]
AppVoyer is happy now, one Travis task fails because of broken PPAs it seems and the other is Android which I'm not sure why this happens.

**Comment by akien-mga** [comment]
```
drivers/gles2/rasterizer_gles2.cpp:176:2: error: use of undeclared identifier 'glDebugMessageControlARB'
        glDebugMessageControlARB(_EXT_DEBUG_SOURCE_API_ARB, _EXT_DEBUG_TYPE_ERROR_ARB, _EXT_DEBUG_SEVERITY_HIGH_ARB, 0, NULL, GL_TRUE);
        ^
```
From a quick googling, it seems like you might need to setup a `ARB_debug_output` context? Or is this extension missing from OpenGL ES 2.0?

File: drivers/gles2/rasterizer_canvas_gles2.cpp [hunk 1]
```diff
@@ -0,0 +1,1145 @@
+/*************************************************************************/
+/*  rasterizer_canvas_gles2.cpp                                          */
+/*************************************************************************/
+/*                       This file is part of:                           */
+/*                           GODOT ENGINE                                */
+/*                      https://godotengine.org                          */
+/*************************************************************************/
+/* Copyright (c) 2007-2017 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2017 Godot Engine contributors (cf. AUTHORS.md)    */
```
Comment (akien-mga) [inline drivers/gles2/rasterizer_canvas_gles2.cpp]: Don't forget to update the years in your copyright headers :)

**Comment by avril-gh** [comment]
:two_hearts:

**Comment by flamendless** [comment]
Thanks for this. I really need this PR since my laptop can only support 2.1 version

**Comment by hpvb** [comment]
I think we should just merge this if there are no build breakages. We're not going to release 3.1 without this, better to let this happen in mainline. 

Also awesome!

**Comment by akien-mga** [comment]
Well, not as ling as CI is broken :)

**Comment by efornara** [comment]
Merging with 3.0.1-stable (https://github.com/efornara/godot - https://github.com/godotengine/godot/issues/16476), I had minor issues with the same files (SConstruct and drivers/SCsub), and they should be trivial to fix.  https://github.com/efornara/godot/commit/a5f100e90683947a92f221aa40af8cb5960f837a

About Android, I suspect the problem is bringing in unavailable definitions.

In my custom platform (https://github.com/efornara/frt), I compile against ES 2.0 directly (i.e. not through glad), and, beside glDebugMessageControlARB, I also had an issue with GL_NUM_EXTENSIONS. https://github.com/efornara/godot/commit/1f62fd35d8be04c568f7914120bcf22090932b92

Once it compiled, I had this problem running the 2d platform demo:

```
ERROR: get_current_version: CanvasShaderGLES2: Vertex shader compilation failed:
0:1(7): error: GLSL 1.20 is not supported. Supported versions are: 1.00 ES, and 3.00 ES

```
I run it on debian stable with this driver:

`OpenGL ES 2.0 Renderer: Mesa DRI Intel(R) Sandybridge Mobile`

Again, I use a custom platform, but my environment should be similar to what you'll find on Android: EGL (no GLX) and GLESv2 (not GL+glad).

**Comment by efornara** [comment]
Well, the run-time error turned out to be easier than I thought to fix: https://github.com/efornara/godot/commit/4e11d8bb31025c3f010265f5dd482e05af8d54f6

By the way, I've just tested the 2d platform demo on a Raspberry Pi and, apart from the lack of particle effects, it works :smiley:

**Comment by akien-mga** [comment]
Boom! Let's get this tested more widely :)

Congrats and thanks for your work @karroffel!

**Comment by groud** [comment]
\\( ﾟヮﾟ)/

**Comment by bruvzg** [comment]
`SConstruct` was rebased incorrectly and revert parts of #16816, #16690 and #16914. Only changes on lines 449-450 should be there.

**Comment by reduz** [comment]
Party!

**Comment by akien-mga** [comment]
> SConstruct was rebased incorrectly and revert parts of #16816, #16690 and #16914. Only changes on lines 449-450 should be there.

Bad, bad @karroffel... :°)

**Comment by karroffel** [comment]
@bruvzg oh yes, I don't know what went wrong there, will make a small fix, sorry

**Comment by bruvzg** [comment]
Which desktop OpenGL version this renderer targets?

I tried to use it on macOS:
 * With `NSOpenGLProfileVersionLegacy` (OpenGL 2.1) context it crashes since OpenGL 3.0+ function [glGenerateMipmap](http://docs.gl/gl4/glGenerateMipmap) is used.
* With `NSOpenGLProfileVersion3_2Core` or `NSOpenGLProfileVersion4_1Core` context shader compilation fails with `version '120' is not supported` error (GLSL 1.20 is from OpenGL 2.1).

macOS does not support OpenGL 3.0 compatibility profile, only 2.1, 3.2 forward-compatible core profile and 4.1 forward-compatible core profile.

**Comment by Oranjoose** [comment]
@efornara That's great you got it working on Raspberry Pi. I however am having a lot of trouble getting v3.0.2 stable to build on Raspberry Pi still. It gives me "KeyError: 'android_arch'". v2.1.3 builds fine.

**Comment by bruvzg** [comment]
@Oranjoose Try disabling WebM module ( add `module_webm_enabled=false` ), there may be something wrong with configuration assembler parts of `libvpx`. I had similar `android_arch` error on Debian/ARM once but OS update fixed it.

**Comment by jtaart** [comment]
Is there a way to test this? I tried a daily build from http://godot3builds.digitecnology.com/, but I still get the opengl 3.3 warning and the engine doesn't start. Do I need to compile it myself with a keyword or something?

**Comment by reduz** [comment]
ah, you will need an ifdef for OpenGL 2.1, since mipmap generation is a
texture parameter

**Comment by bruvzg** [comment]
> Is there a way to test this?

@jtaart Add ` --video-driver GLES2 ` to command line.
It should work with OpenGL 2.1 + `GL_ARB_framebuffer_object` or `GL_EXT_framebuffer_object`.

**Comment by jtaart** [comment]
@bruvzg Thanks, running with --video-driver GLES2 in command prompt worked to get into project select, but when I selected the project it would give me the warning again. Then I found a way to run straight to the project and get into the editor. The problem I have now is I can't run the game from the editor without the the warning. I think I'm not fully understanding how to do it. 

https://i.imgur.com/QyMdYZk.gifv

**Comment by bruvzg** [comment]
@jtaart When you open a project from project manager it's starting a new copy of Godot and command line arguments are not passed to it. You have to open project from command line with `--video-driver GLES2 --path PATH_TO_PROJECT_FOLDER --editor` for the first time and change default driver in `Project Settings` -> `Rendering` -> `Quality` -> `Driver Name` -> `GLES2`. Next time you should be able to open this project directly from Project Manager.

**Comment by rraallvv** [comment]
@karroffel it seems the option **Driver Name** affects both the editor and the game at runtime, do you think it's a good idea to have separate options for each one? So, for instance one could run the editor with full GLES3 capability, and then run the game or deploy to the device with only the GLES2 diver enabled. Maybe a way to detect GLES3 and fall back to GLES2 if not available would come in handy too.

**Comment by starry-abyss** [comment]
I believe it's not possible to detect all buggy drivers' GLES3 (un)support, but maybe it makes sense to default GLES2 for the editor when it's ready?

**Pull request merged**